geode-core/src/main/java/org/apache/geode/internal/cache/TXCommitMessage.java

```diff
@@ -391,9 +404,16 @@ void send(TXLockId lockId) {
                 setRecipientsSendData(Collections.singleton(indivRecip.next()), processor, rcl);
               }
             } else {
-              // Run in normal mode sending to multiple recipients in
-              // one shot
-              setRecipientsSendData(recipients, processor, rcl);
+              if (this.notificationOnlyMembers.isEmpty()) {
+                // Run in normal mode sending to multiple recipients in one shot
+                setRecipientsSendData(recipients, processor, rcl);
+              } else {
+                recipients.removeAll(this.notificationOnlyMembers);
+                setRecipientsSendData(recipients, processor, rcl);
+
+                this.txState.setTailKeyOnEntries(-1L);
+                setRecipientsSendData(notificationOnlyMembers, processor, rcl);
```

Comment:
Yes this makes sense. In this case If I understood correctly we would have to extend TXCommitMessage message with notificationOnly list of servers. This seems like better solution, and I think it is manageable. Also I have posted comment in  https://github.com/apache/geode/pull/6477#issuecomment-848723153 another alternative which would require impact only on receiver side (adjunct member). Could you please take a look and comment?

Reply:
Hi @DonalEvans ,

Not sure that I completely understand your question:

"Since we're now splitting the sending of this message up into two separate calls to DistributionManager.putOutgoing() might there be problems if the first call succeeds but the second doesn't?"

What I could see from code and some tests is that this `DistributionManager.putOutgoing()` is executed asynchronously. This means that messages are first sent to the receivers and then replies are checked later on with `CommitReplyProcessor` processor. `CommitReplyProcessor` will wait for replies from all receivers (used in first and second call of DistributionManager.putOutgoing()). If any commit message fail (for example second call of `DistributionManager.putOutgoing()`) this will be recorded in processor and `org.apache.geode.cache.CommitIncompleteException` will be returned to client.

I have tested both that exception only occurs on member that host secondary bucket (first call of `DistributionManager.putOutgoing()`) or only on member that host only CQ queue(second call of `DistributionManager.putOutgoing()`). Same exception `CommitIncompleteException` propagated to client.

Not sure if you had something else in mind?

Reply:
After discussing the changes with some more knowledgeable people, I now realise that there's no cause for concern here at all, since the `TXCommitMessage` is just the first part of applying the transaction commit, and there won't be any inconsistency if one or all of the messages fail, because no `CommitProcessForTXIdMessage` will be sent in that case. I'm fine with this section as it is.